src/Control/Arrow/Free.hs

```diff
@@ -0,0 +1,210 @@
+-- {-# LANGUAGE AllowAmbiguousTypes   #-}
+{-# LANGUAGE Arrows                #-}
+{-# LANGUAGE ConstraintKinds       #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE InstanceSigs          #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NoImplicitPrelude     #-}
+{-# LANGUAGE PolyKinds             #-}
+{-# LANGUAGE Rank2Types            #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE TypeOperators         #-}
+
+module Control.Arrow.Free
+  ( Free
+  , Choice
+  , ErrorChoice
+  , effect
+  , eval
+    -- * ArrowError
+  , ArrowError(..)
+    -- * Arrow functions
+  , mapA
+  , filterA
+  ) where
+
+import           Control.Arrow
+import           Control.Category
+import           Control.Monad.Catch (Exception, MonadCatch)
+import qualified Control.Monad.Catch as Monad.Catch
+import           Data.Bool           (Bool)
+import           Data.Constraint     (Constraint, Dict (..), mapDict, weaken1,
+                                      weaken2)
+import           Data.Either         (Either (..))
+import           Data.Function       (const, flip, ($))
+import           Data.List           (uncons)
+import           Data.Maybe          (maybe)
+import           Data.Tuple          (curry, uncurry)
+
+type x ~> y = forall a b. x a b -> y a b
+
+--------------------------------------------------------------------------------
+-- FreeArrowLike
+--------------------------------------------------------------------------------
+
+-- | Small class letting us define `eval` and `effect` generally over
+--   multiple free structures
+class FreeArrowLike fal where
```

Comment:
Everything would be so much simpler (and without "annoying hackery") if we didnt have this class, just have `evalFree`, `evalChoice` etc